accord-core/src/main/java/accord/utils/async/AsyncResults.java

```diff
@@ -0,0 +1,329 @@
+            {
+                Object current = state;
+                if (current instanceof Result)
+                    return false;
+                Listener<V> listener = (Listener<V>) current;
+                if (STATE.compareAndSet(this, current, result))
+                {
+                    notify(listener, result);
+                    return true;
+                }
+            }
+        }
+
+        boolean trySetResult(V result, Throwable failure)
+        {
+            return trySetResult(new Result<>(result, failure));
+        }
+
+        private  AsyncChain<V> newChain()
+        {
+            return new AsyncChains.Head<V>()
+            {
+                @Override
+                protected void start(BiConsumer<? super V, Throwable> callback)
+                {
+                    AbstractResult.this.addCallback(callback);
+                }
+            };
+        }
+
+        void setResult(V result, Throwable failure)
+        {
+            if (!trySetResult(result, failure))
+                throw new IllegalStateException("Result has already been set on " + this);
+        }
+
+        @Override
+        public <T> AsyncChain<T> map(Function<? super V, ? extends T> mapper)
+        {
+            return newChain().map(mapper);
+        }
+
+        @Override
+        public <T> AsyncChain<T> flatMap(Function<? super V, ? extends AsyncChain<T>> mapper)
+        {
+            return newChain().flatMap(mapper);
+        }
+
+        @Override
+        public AsyncResult<V> addCallback(BiConsumer<? super V, Throwable> callback)
+        {
+            Listener<V> listener = null;
+            while (true)
+            {
+                Object current = state;
+                if (current instanceof Result)
+                {
+                    Result<V> result = (Result<V>) current;
+                    callback.accept(result.value, result.failure);
+                    return this;
+                }
+                if (listener == null)
+                    listener = new Listener<>(callback);
+
+                listener.next = (Listener<V>) current;
+                if (STATE.compareAndSet(this, current, listener))
+                    return this;
+            }
+        }
+
+        @Override
+        public boolean isDone()
+        {
+            return state instanceof Result;
+        }
+
+        @Override
+        public boolean isSuccess()
+        {
+            Object current = state;
+            return current instanceof Result && ((Result) current).failure == null;
+        }
+
+        private Result<V> getResult()
+        {
+            Object current = state;
+            Invariants.checkState(current instanceof Result);
+            return (Result<V>) current;
+        }
+
+        public V result()
+        {
+            Result<V> result = getResult();
+            if (result.failure != null)
+                throw new IllegalStateException("Result failed " + result.failure);
```

Comment:
```suggestion
                throw new IllegalStateException("Result failed", result.failure);
```

so we don't loose the stack trace